libnavigation-ui/src/main/java/com/mapbox/navigation/ui/NavigationViewModel.java

```diff
@@ -396,15 +396,17 @@ private SpeechPlayerProvider initializeSpeechPlayerProvider(boolean voiceLanguag
   }
 
   private LocationEngine initializeLocationEngineFrom(final NavigationViewOptions options) {
-    if (options.shouldSimulateRoute()) {
+    if (options.locationEngine() != null) {
+      return options.locationEngine();
+    } else if (options.shouldSimulateRoute()) {
       ReplayLocationEngine replayLocationEngine = new ReplayLocationEngine(mapboxReplayer);
       final Point lastLocation = getOriginOfRoute(options.directionsRoute());
       ReplayEventBase replayEventOrigin = ReplayRouteMapper.mapToUpdateLocation(0.0, lastLocation);
       mapboxReplayer.pushEvents(Collections.singletonList(replayEventOrigin));
       mapboxReplayer.play();
       return replayLocationEngine;
     } else {
-      return options.locationEngine();
+      return navigation.getLocationEngine();
```

Comment:
Isn't this going to end up in the null pointer exception? The `navigation` field is not yet initialized when `initializeLocationEngineFrom` is called.

Reply:
this location engine is initialized here https://github.com/mapbox/mapbox-navigation-android/blob/d7b7dcb2e4be364c7047a1737eb189c4bed76640/libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt#L125

it can't be null so, i don't think that could be an issue.

The issues are when there are different instances of the LocationEngine. It's possible to create multiple LocationEngines and give them to different components. Do we want to support that?

Reply:
> The navigation field is not yet initialized when initializeLocationEngineFrom is called

oh wait, `navigation`, oh yeah java.

Reply:
Interesting it didn't crash for the examples for me. I think it's because the call stack is coming from `NavigationView.startNavigation`

It will crash if you are able to call startNavigation before initializing MapboxNavigation. Which is what it should be no?

Reply:
```
      LocationEngine locationEngine = initializeLocationEngineFrom(options);
      initializeNavigation(getApplication(), updatedOptions, locationEngine);
```

Reply:
Thanks for the catch @LukasPaczos https://github.com/mapbox/mapbox-navigation-android/pull/3174